<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/Python-Notebook-Banners/Exercise.png"  style="display: block; margin-left: auto; margin-right: auto;";/>
</div>

# Exercise: Principal component analysis
© ALX

In this exercise, we apply PCA to a dataset, evaluate the cumulative variance explained, and determine the appropriate number of components to retain.

## Learning objectives

By the end of this notebook, you should be able to:
* Apply PCA to reduce a dataset’s dimensionality.
* Evaluate the cumulative variance explained by each principal component.
* Determine the number of components needed to capture at least 85% of the variance. 

## Overview

The Digits dataset consists of 1,797 images of handwritten digits, each represented by a 64-dimensional feature vector. The dataset's high dimensionality can pose challenges when visualising and exploring it and could also lead to model complexity.

In this exercise, we apply PCA to the Digits dataset and evaluate its ability to reduce the dataset's dimensionality while retaining valuable information.

## Import libraries 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_digits
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

## Load and prepare dataset

In [2]:
# Load the dataset
digits = load_digits()
X = digits.data
y = digits.target

# Standardize the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [10]:
X

array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

## Exercises

### Exercise 1

To reduce the dataset's dimensionality, let's transform the standardised dataset by applying PCA.

In [7]:
pca = PCA(.85)

pca_X_scaled_85 = pca.fit_transform(X_scaled)

pca_explain_variance_ratio = pca.explained_variance_ratio_.sum()*100
print("explained_variance_ratio {}".format(pca_explain_variance_ratio))
num_components = pca.n_components_
print("number of component explain in 85%: {}".format(num_components))

explained_variance_ratio 85.13246353478716
number of component explain in 85%: 25


### Exercise 2

To understand which components carry the most information, we can assess how much of the dataset's variance is captured by each principal component.

Compute and print the `Explained Variance Ratio` for each principal component formatted to four decimal places.

In [14]:
for col, val in enumerate(pca.explained_variance_ratio_):
    print(f"Principal Component {col + 1} with a variance of {val * 100:.2f}%")


Principal Component 1 with a variance of 12.03%
Principal Component 2 with a variance of 9.56%
Principal Component 3 with a variance of 8.44%
Principal Component 4 with a variance of 6.50%
Principal Component 5 with a variance of 4.86%
Principal Component 6 with a variance of 4.21%
Principal Component 7 with a variance of 3.94%
Principal Component 8 with a variance of 3.39%
Principal Component 9 with a variance of 3.00%
Principal Component 10 with a variance of 2.93%
Principal Component 11 with a variance of 2.78%
Principal Component 12 with a variance of 2.58%
Principal Component 13 with a variance of 2.28%
Principal Component 14 with a variance of 2.23%
Principal Component 15 with a variance of 2.17%
Principal Component 16 with a variance of 1.91%
Principal Component 17 with a variance of 1.78%
Principal Component 18 with a variance of 1.64%
Principal Component 19 with a variance of 1.60%
Principal Component 20 with a variance of 1.49%
Principal Component 21 with a variance of 1.35%


### Exercise 3

We can also evaluate how much total variance is captured as components are added incrementally. This can help us get a view of how many components are needed to capture a substantial proportion of the dataset's variance.

Determine the cumulative variance ratio by summing the explained variance ratios of each principal component.

In [16]:
total_variance = pca.explained_variance_ratio_.sum() * 100
print(f"total cummulative variance {total_variance}")

total cummulative variance 85.13246353478716


### Exercise 4

Based on the results from **Exercise 3**, determine how many components are needed to capture at least 85% of the total variance.

Discuss the impact of this on subsequent analysis or modeling.

In [17]:
pca.n_components_

25

## Solutions

### Exercise 1

In [18]:
# Apply PCA
pca = PCA()
X_pca = pca.fit_transform(X_scaled)

### Exercise 2

In [11]:
# Get the explained variance ratio
explained_variance_ratio = pca.explained_variance_ratio_

# Print the explained variance ratio
for i, ev in enumerate(explained_variance_ratio):
    print(f"PC{i+1}: Explained Variance = {ev:.4f}")

PC1: Explained Variance = 0.1203
PC2: Explained Variance = 0.0956
PC3: Explained Variance = 0.0844
PC4: Explained Variance = 0.0650
PC5: Explained Variance = 0.0486
PC6: Explained Variance = 0.0421
PC7: Explained Variance = 0.0394
PC8: Explained Variance = 0.0339
PC9: Explained Variance = 0.0300
PC10: Explained Variance = 0.0293
PC11: Explained Variance = 0.0278
PC12: Explained Variance = 0.0258
PC13: Explained Variance = 0.0228
PC14: Explained Variance = 0.0223
PC15: Explained Variance = 0.0217
PC16: Explained Variance = 0.0191
PC17: Explained Variance = 0.0178
PC18: Explained Variance = 0.0164
PC19: Explained Variance = 0.0160
PC20: Explained Variance = 0.0149
PC21: Explained Variance = 0.0135
PC22: Explained Variance = 0.0127
PC23: Explained Variance = 0.0117
PC24: Explained Variance = 0.0106
PC25: Explained Variance = 0.0098


The explained variance ratios show that the first few principal components (PC1, PC2, PC3) capture the most variance, with PC1 accounting for 12.03%. As we move to higher components, the explained variance decreases, indicating they carry less information about the dataset's overall variance.

### Exercise 3

In [ ]:
# Calculate the cumulative variance ratio
cumulative_variance_ratio = np.cumsum(explained_variance_ratio)

# Print the cumulative variance ratio
for i, cv in enumerate(cumulative_variance_ratio):
    print(f"PC{i+1}: Cumulative Variance = {cv:.4f}")

The results show that each additional principal component captures more variance. For example, PC1 captures 12.03%, and by PC10, 58.87% is captured. All 64 components capture 100% variance. This helps identify how many components are needed to capture a significant portion of the dataset's variance.

### Exercise 4

The cumulative variance ratio shows that to reach at least 85% cumulative variance, 25 components are needed.

We observe that by retaining the first 25 components, we can capture 85.13% of the total variance in the dataset, which significantly reduces the dataset's dimensionality while retaining most of its information.

Using 25 components instead of the original 64 simplifies any downstream models by reducing their feature space, potentially improving model performance and interpretability.

The reduced number of components also makes visualising the data easier, which can provide meaningful insights into class separations or clustering.

Therefore, the ability to capture over 85% of the variance with 25 components makes PCA a viable dimensionality reduction technique for this dataset, preserving most information while simplifying further analyses.

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/refs/heads/master/ALX_banners/ALX_Navy.png"  style="width:100px";/>
</div>